include/ginkgo/core/base/block_operator.hpp

```diff
@@ -0,0 +1,197 @@
+#ifndef GINKGO_BLOCK_MATRIX_HPP
+#define GINKGO_BLOCK_MATRIX_HPP
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+
+
+namespace gko {
+namespace detail {
+
+
+/**
+ * Carbon copy og gko::span, except that the members are not const.
+ * Necessary since the normal gko::span is not copy/movable.
+ */
+struct value_span {
+    constexpr value_span(size_type point) noexcept
+        : value_span{point, point + 1}
+    {}
+
+    constexpr value_span(size_type begin, size_type end) noexcept
+        : begin{begin}, end{end}
+    {}
+
+    constexpr operator span() const { return {begin, end}; }
+
+    constexpr value_span(const span& s) noexcept : value_span(s.begin, s.end) {}
+
+    constexpr bool is_valid() const { return begin <= end; }
+
+    constexpr size_type length() const { return end - begin; }
+
+    size_type begin;
+    size_type end;
+};
+
+
+}  // namespace detail
+
+
+/**
+ * A BlockOperator represents a linear operator that is partitioned into
+ * multiple blocks.
+ *
+ * For example, a BlockOperator can be used to define the operator:
+ * ```
+ * | A B |
+ * | C D |
+ * ```
+ * where `A, B, C, D` itself are matrices of compatible size.
+ * The requirements on the individual blocks are:
+ * - In each block-row, all blocks have the same number of rows
+ * - In each block-column, all blocks have the same number of columns
+ * - Each block-row must have the same number of blocks
+ * It is possible to set blocks to zero, by passing in a nullptr. But every
+ * block-row and block-column must contain at least one non-nullptr block.
+ *
+ * The constructor will store all passed in blocks on the same executor as the
+ * BlockOperator, which will requires copying any block that is associated with
+ * a different executor.
+ */
+class BlockOperator final : public EnableLinOp<BlockOperator> {
+    friend class EnablePolymorphicObject<BlockOperator, LinOp>;
+
+public:
+    /**
+     * Get the block dimension of this, i.e. the number of blocks per row and
+     * column.
+     *
+     * @return  The block dimension of this.
+     */
+    dim<2> get_block_size() const { return block_size_; }
+
+    /**
+     * Const access to a specific block.
+     *
+     * @param i  block row.
+     * @param j  block column.
+     *
+     * @return  the block stored at (i, j).
+     */
+    const LinOp* block_at(size_type i, size_type j) const
+    {
+        return blocks_[i * block_size_[1] + j].get();
+    }
+
+    /**
+     * Mutable acces to a specific block.
+     *
+     * @copydoc block_at(size_type, size_type)
+     */
+    LinOp* block_at(size_type i, size_type j)
+    {
+        return blocks_[i * block_size_[1] + j].get();
+    }
+
+    /**
+     * Create empty BlockOperator.
+     *
+     * @param exec  the executor of this.
+     *
+     * @return  empty BlockOperator.
+     */
+    static std::unique_ptr<BlockOperator> create(
+        std::shared_ptr<const Executor> exec);
```

Comment:
It would need both block size and the combined size of the blocks to make sense. But a constructor based on those would rely on users filling the blocks after the block operator has been created. That gives too much room for error, a user could easily set a block to the wrong sizes. So I would prefer not allowing users to change the blocks after the block operator has been created. (I will also remove the mutable block access)